### DISCUSSION TITLE
Open Telemetry dropping child spans between Kafka Connect and Spring KafkaListener Consumer

### DISCUSSION THREAD

**Discussion by vishnusaikumar** [open-telemetry/opentelemetry-java-instrumentation#8126] [Q&A]
We have DockerFile which uses Jaeger all in one image

![image](https://user-images.githubusercontent.com/16996345/227489534-1c127cc4-aa28-41c4-ace7-e6c40eee7d27.png)

We have enabled OpenTelemetry java agent with Kafka Agent as shown below

![image](https://user-images.githubusercontent.com/16996345/227489060-d11a2a49-154b-4fce-9723-59598cb03199.png)

At Kafka connect producer topic, the count (No. of records at producer topic shown in Akhq tool) is matching with below aggregated elastic search query based on Trace IDs

![image](https://user-images.githubusercontent.com/16996345/227490328-b4c86066-d8b8-4d76-9c0c-8a8d7efbf236.png)

Now at consumer we have a configured similarly

![image](https://user-images.githubusercontent.com/16996345/227491422-c022e3c4-0cd6-407a-a59a-9428f5ea55b5.png)

We have a Kafka Listener at consumer which listens to messages in topic

```
@KafkaListener(topics ="${spring.kafka.properties.topics.qualityGaps}", containerFactory = "mbrQualityKafkaListenerContainerFactory" )
	public void onMessage(ConsumerRecord<String, MbrQuality> record, Acknowledgment acknowledgment) {
		List<ConsumerRecord<String, MbrQuality>> records = new ArrayList<>();
		records.add(record);
		try {
			if (Objects.nonNull(record) && !records.isEmpty()) {
				int recordsReceived = records.size();
				log.info("Received {} memberQuality records from DataSync.MemberGaps.Quality", recordsReceived);
				long start = System.currentTimeMillis();
				List<MbrQuality> batch = records.stream().map(ConsumerRecord::value).collect(Collectors.toList());
				log.info("Processing records of partition {} from offset {}",records.get(0).partition(), records.get(0).offset());
				int rowsAffected =mbrQualityService.process(batch);
				records.clear();
				acknowledgment.acknowledge();
				long end = System.currentTimeMillis();
				log.info("Received records: {}; Persisted MemberQuality records: {} in {} seconds", recordsReceived, rowsAffected, (end-start)/1000.0);
			}

		} catch (Exception exception) {
			List<Object> recordList = new ArrayList<>();
			records.stream().map(ConsumerRecord::value).collect(Collectors.toList()).
					forEach(mbrQuality -> {
						if(!Objects.isNull(mbrQuality.getAfter())) {
							recordList.add(
									mbrQuality.getAfter().get("MBR_QUALITY_SK"));
						}
					});
			log.error("Error while processing MemberQualityGaps, Failed Records List:{} with exception ", recordList, exception);
			acknowledgment.acknowledge();
			boolean publishToRetryStatus = mbrQualityService.publishToRetryDLTTopic(records, true);
			log.info ("Publish to MemberQuality Retry Topic is {}", publishToRetryStatus);
		}
	}
```

Also added some interceptors at Kafka Container Factory

`consumerProps.put(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG, Collections.singletonList(TracingConsumerInterceptor.class));`

Now 
the number of spans at consumer is always less than at producer topic even though the consumption is happening for all the records 

![image](https://user-images.githubusercontent.com/16996345/227492965-0df39b85-03cf-4ad9-8b4f-e46f95c5284b.png)

Multiple child spans are getting dropped

We are using 

Spring Boot 2.3
and
Spring Kafka 2.5.1 Release

Is there any additional Configuration needed or any limitations from Kafka Instrumentation?

Any suggestions would help
Thanks

**Comment by vishnusaikumar** [comment]
Suspecting on Batch-Listener we have disabled batchListener and made max poll records to 1

     ```
 concurrentKafkaListenerContainerFactory.batch.listener: false
      qualityGaps.concurrentKafkaListenerContainerFactory.concurrency: 1000
      qualityGaps.max.poll.records: 1
```

but we don't see any improvement in child traces

**Comment by vishnusaikumar** [comment]
Followed 

[https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/7033](url)

and added 


```
protected <K, V> KafkaListenerContainerFactory<ConcurrentMessageListenerContainer<K, V>> createContainerFactory(
            String topicName, Deserializer<K> keyDeserializer, Deserializer<V> valueDeserializer, String consumer) {

ConcurrentKafkaListenerContainerFactory<K, V> factory = new ConcurrentKafkaListenerContainerFactory<> ();

SpringKafkaTelemetry springKafkaTelemetry = SpringKafkaTelemetry.create(GlobalOpenTelemetry.get());

factory.setRecordInterceptor(springKafkaTelemetry.createRecordInterceptor());

factory.setBatchInterceptor(springKafkaTelemetry.createBatchInterceptor());

}
```

and some other properties as well


Upgraded to 
Spring Boot 2.7.5
Spring Kafka 2.8.10
and using 

`<dependency>
            <groupId>io.opentelemetry.instrumentation</groupId>
            <artifactId>opentelemetry-spring-kafka-2.7</artifactId>
            <version>1.24.0-alpha</version>
        </dependency>`

for above SpringKafkaTelemetry import

Still I see a lot of difference in traces count at Consumer end

Please suggest any solutions here 
Thanks

**Comment by vishnusaikumar** [comment]
Is this issue related to  
[https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2623](url)

Because we Configured OpenTelemetry Java agent at Kafka Connect project and traces are getting generated from Kafka Connect only and also configured java agent at Spring Kafka Consumer

**Comment by vishnusaikumar** [comment]
Can anyone suggest a solution to this yet?

**Comment by laurit** [comment]
Can you create a small sample application that reproduces the issue? Or is this something that happens only when you send/receive a lot of messages?

**Comment by vishnusaikumar** [comment]
Irrespective of the number of messages sent/received the child spans are always getting dropped.
As I mentioned in the first chat 
"At Kafka connect producer topic, the count (No. of records at producer topic shown in Akhq tool) is matching with below aggregated elastic search query based on Trace IDs"
but the issue is at Kafka listener at Consumer where trace ID propagation is not happening

We have checked with 10000 records at Kafka Connect for which around 4000 child records are only matching at Kafka listener in consumer (stage environment)

In production, this ratio always fluctuates between 20 to 40 percent with around 10 million records at Kafka Connect on daily basis.

For the sample code, I have given most of the screenshots in the chat earlier. Let me know if you need any additional information because the sample code might take some time to reproduce.

**Reply by laurit** [reply]
You are using `-Dotel.instrumentation.messaging.experimental.receive-telemetry.enabled=true` This flag tells the instrumentation that producer and consumer spans should be in separate traces that are connected via a span link. You are also using `-Dotel.traces.exporter=zipkin` Zipkin trace format does not support span links. From this I'd say it is a bit surprising that you see send and receive spans in the same trace at all.
If you believe that this is a bug and you wish someone to take a look at this you have to provide a minimal application that reproduces this issue. It is unlikely that you'll find someone here who'd be willing to attempt to reproduce this issue based on the code snippets you provided.

**Comment by vishnusaikumar** [comment]
Thank you for your response
So you are saying **"Zipkin trace format does not support span links."**
Is there any alternative like using **Jaeger exporter** ?
I will try to add a minimal application and link it here soon.

**Reply by laurit** [reply]
Yes, jaeger exporter supports span links.
